sled-agent/src/bootstrap/server.rs

```diff
@@ -58,114 +65,183 @@ impl Server {
+    log: Logger,
+}
+
+impl Inner {
+    async fn start(
+        bind_address: SocketAddrV6,
+        sp: Option<SpHandle>,
+        bootstrap_agent: Arc<Agent>,
+        log: Logger,
+    ) -> Result<JoinHandle<Result<(), String>>, String> {
+        let listener =
+            TcpListener::bind(bind_address).await.map_err(|err| {
+                format!("could not bind to {bind_address}: {err}")
+            })?;
+        info!(log, "Started listening"; "local_addr" => %bind_address);
+        let inner = Inner { listener, sp, bootstrap_agent, log };
+        Ok(tokio::spawn(inner.run()))
+    }
+
+    // Run our sprockets server. The only `.await` point is when `accept()`ing
+    // on our bound socket, so we can be cleanly shut down by our caller
+    // `.abort()`ing our task, which will not affect any already-accepted
+    // sockets (which are spawned onto detached tokio tasks).
+    async fn run(self) -> Result<(), String> {
+        loop {
+            let (stream, remote_addr) =
+                self.listener.accept().await.map_err(|err| {
+                    format!("accept() on already-bound socket failed: {err}")
+                })?;
+
+            let log = self.log.new(o!("remote_addr" => remote_addr));
+            info!(log, "Accepted connection");
+
+            let sp = self.sp.clone();
+            let ba = Arc::clone(&self.bootstrap_agent);
+            tokio::spawn(async move {
+                match serve_single_request(stream, sp, ba, &log).await {
+                    Ok(()) => info!(log, "Connection closed"),
+                    Err(err) => warn!(log, "Connection failed"; "err" => err),
+                }
+            });
+        }
+    }
 }
 
-async fn spawn_sprockets_proxy(
-    sp: &Option<SpHandle>,
-    http_server: &HttpServer<Arc<Agent>>,
-    sprockets_proxy_bind_addr: SocketAddrV6,
+async fn serve_single_request(
+    stream: TcpStream,
+    sp: Option<SpHandle>,
+    bootstrap_agent: Arc<Agent>,
     log: &Logger,
 ) -> Result<(), String> {
-    // We can only start a sprockets proxy if we have an SP.
-    let sp = sp.as_ref().ok_or(
-        "Misconfiguration: cannot start a sprockets proxy without an SP",
-    )?;
-
-    // If we're running a sprockets proxy, our dropshot server should be
-    // listening on localhost.
-    let dropshot_addr = http_server.local_addr();
-    if !dropshot_addr.ip().is_loopback() {
-        return Err(concat!(
-            "Misconfiguration: bootstrap dropshot IP address should ",
-            "be loopback when using a sprockets proxy"
-        )
-        .into());
+    // Bound to avoid allocating an unreasonable amount of memory from a bogus
+    // length prefix from a client. We authenticate clients via sprockets before
+    // allocating based on the length prefix they send, so it should be fine to
+    // be a little sloppy here and just pick something far larger than we ever
+    // expect to see.
+    const MAX_REQUEST_LEN: u32 = 128 << 20;
```

Comment:
Maybe share this between client and server?

Reply:
I quasi-intentionally picked different sizes here, although maybe that was being too clever. Requests will certainly be larger than responses in general (especially as the init request grows to include the list of devices in the quorum), but in practice we don't expect this bound to ever be hit in either case. I have a slight preference for keeping them separate only because it allows them to be local to their respective functions, but if you'd rather I share one I'm fine with that.

Reply:
Oh, I didn't even notice they were different. I'm fine with this as is. Carry on!